libraries/AP_HAL/CAN.h

```diff
@@ -0,0 +1,139 @@
+ *
+ * This file is distributed in the hope that it will be useful, but
+ * WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#pragma once
+
+#include <inttypes.h>
+#include <string.h>
+#include <assert.h>
+#include <cmath>
+
+#include "AP_HAL_Namespace.h"
+#include "utility/functor.h"
+#include <uavcan/driver/can.hpp>
+#include <uavcan/time.hpp>
+
+class AP_UAVCAN;
+
+/**
+ * Single non-blocking CAN interface.
+ */
+class AP_HAL::CAN: public uavcan::ICanIface {
+public:
+    /*  CAN port open method
+
+     Opens port with specified bit rate and type.
+     CAN 2.0B is back compatible with CAN 2.0A and such devices can be opened with both types
+
+     bitrate     Selects the speed that the port will be configured to.  If zero, the port speed is left
+     unchanged.
+
+     return      0 - CAN port open failed
+     code of error can be read with get_last_error() method
+     1 - CAN port open succeeded
+     */
+    virtual bool begin(uint32_t bitrate) = 0;
+
+    /*
+     CAN port close
+
+     return      0 - CAN port close failed
+     code of error can be read with get_last_error() method
+     1 - CAN port close succeeded
+     */
+    virtual void end() = 0;
+
+    /*
+     Reset opened CAN port
+
+     Pending messages to be transmitted are deleted and receive state and FIFO also reset.
+     All pending errors are cleared.
+     */
+    virtual void reset() = 0;
+
+    /*
+     Test if CAN port is opened and initialized
+
+     return      0 - CAN port not initialized
+     code of error can be read with get_last_error() method
+     1 - CAN port is initialized
+     */
+    virtual bool is_initialized() = 0;
+
+    /*
+     Return if CAN port has some untransmitted pending messages
+
+     return     -1 - CAN port is not opened or initialized
+     0 - no messages are pending
+     positive - number of pending messages
+     */
+    virtual int32_t tx_pending() = 0;
+
+    /*
+     Return if CAN port has received but yet read messages
+
+     return     -1 - CAN port is not opened or initialized
+     0 - no messages are pending for read
+     positive - number of pending messages for read
+     */
+    virtual int32_t available() = 0;
+
+};
+
+/**
+ * Generic CAN driver.
+ */
+class AP_HAL::CANManager: public uavcan::ICanDriver {
+public:
+    /**
+     * Returns an interface by index, or null pointer if the index is out of range.
+     */
+    virtual AP_HAL::CAN* getIface(uint8_t iface_index) = 0;
+
+    /**
+     * Default implementation of this method calls the non-const overload of getIface().
+     * Can be overriden by the application if necessary.
+     */
+    virtual const AP_HAL::CAN* getIface(uint8_t iface_index) const
```

Comment:
> How is that safe? You don't control the implementation, there's no way to know if you are breaking the constness contract or not.
> But why is this second method const if the first isn't? Why does this second method exist?

Observe that they return different values. The non-const method returns a pointer to non-const instance of `ICanIface`, whereas the const method returns a pointer to a const instance of `ICanIface`. If this helper breaks the const contracts in the implementation (although I can't imagine a sensible use case where this could happen), the developer of the implementation will have to define both methods manually.

That said, I am perfectly fine with requiring the implementation to always provide both methods.